src/foam/graphics/TreeNode.js

```diff
@@ -0,0 +1,297 @@
+        var nodeLeftEdgePadded = x - nodeWidth / 2 - padding / 2;
+        var nodeRightEdgePadded = x + nodeWidth / 2 + padding / 2;
+        var rootLevelOutline = [
+          {
+            left: nodeLeftEdgePadded,
+            right: nodeRightEdgePadded
+          }
+        ];
+
+        var champion = [];
+
+        for (let i = 0; i < childNodes.length && expanded; i++) {
+          // get child outline
+          // transform all rows
+          // merge levels into childoutlines
+          var childOutline = childNodes[i].outline.map(o => ({
+            left: o.left + x,
+            right: o.right + x,
+          }));
+
+          for (var j = 0; j < childOutline.length; j++) {
+            champion[j] = champion[j] || {
+              left: Number.MAX_SAFE_INTEGER,
+              right: Number.MIN_SAFE_INTEGER
+            };
+            champion[j].left = Math.min(childOutline[j].left, champion[j].left);
+            champion[j].right = Math.max(childOutline[j].right, champion[j].right);
+          }
+        }
+
+        var totalOutline = rootLevelOutline.concat(champion);
+
+        return totalOutline;
+      }
+    },
+    { name: 'height', factory: function () { return this.nodeHeight; } },
+    { name: 'width', factory: function () { return this.nodeWidth; } },
+    ['border', 'gray'],
+    {
+      name: 'childNodes',
+      factory: function () { return []; }
+    },
+    ['left', 0],
+    ['right', 0],
+    ['maxLeft', 0],
+    ['maxRight', 0],
+    {
+      name: 'x',
+      expression: function(neighborX, centerX) {
+        return neighborX + centerX;
+      },
+      postSet: function() {
+        this.neighborX = 0;
+        this.centerX = 0;
+        this.x = undefined;
+      }
+    },
+    {
+      class: 'Float',
+      name: 'neighborX'
+    },
+    {
+      class: 'Float',
+      name: 'centerX'
+    },
+    {
+      class: 'Boolean',
+      name: 'expanded',
+      expression: function(childNodes) {
+        return childNodes.length < 5;
+      },
+      postSet: function() {
+        this.graph.doLayout();
+      }
+    },
+    ['color', 'white']
+  ],
+
+  methods: [
+    function initCView() {
+      this.SUPER();
+
+      this.formatNode();
+
+      if (this.relationship) {
+        var data = this.data.clone(this.__subContext__);
+        try {
+          data[this.relationship.forwardName].select(data => {
+            this.addChildNode({ data: data });
+          });
+        } catch (x) { }
+        this.graph.doLayout();
+      }
+
+      this.canvas.on('click', (e) => {
+        var p = this.parent;
+
+        while ( this.cls_.isInstance(p) ) {
+          if ( ! p.expanded ) return;
+          p = p.parent;
+        }
+
+        var point = this.globalToLocalCoordinates({
+          w: 1,
+          x: e.offsetX,
+          y: e.offsetY,
+        });
+        point.x += this.width / 2;
+
+        var hit = this.hitTest(point)
```

Comment:
Do you need to store this in a variable or can you use it directly in the if statement?

Reply:
Yeah its not used anywhere else, moved it directly in the if statement